### PR TITLE
Enable test mode during prediction

### PR DESCRIPTION
**Pull request by Yangqing**
This is similar to @bwasti 's pull request #7 : during testing, Caffe/Caffe2 passes an "is_test" option to remove the dropout effect. This can be achieved by passing the arg scope to the model constructor.

With this I am recording a 79% accuracy, which makes me wonder if there is a minor discrepancy in the underlying training code - I was expecting 78%, but it might be due to randomness.

### CODE REVIEW

**Comment by Yangqing** [comment]
Thanks for the quick merge! If you could rerun the notebook that would be great, I tried to reserve an Azure gpu instance but they are sold out.

**Comment by ilkarman** [comment]
Yangqing thanks for both of the changes! I have re-run on the K80 and both accuracy and speed are top now! I'm curious how this went from 75 to 79 whereas others went from 72 to 77 (dropout test-flag) and will check the code. The speed bump from not computing gradient for the first input gave a nice speed bump, curious why this isn't default

**Comment by ilkarman** [comment]
Amazing speed given that I feel I'm almost abusing how it should be trained by using a custom iterator with a numpy array instead of the database loader

**Comment by Yangqing** [comment]
Thanks for rerunning it - yeah, a few things are bordering tricks so we did not enable it by default. Here is a detailed explanation in case it is useful in general:

(1) Not computing gradients for input: this is a minor thing for very large nets, and for research purposes we often want to inspect the gradient wrt the input. It is also useful for networks like deep image, so we keep it in default. This also matches the conventional protocol when benchmarking.

(2) Accuracy - I honestly don't know, and I think it's probably because of initialization. Took a look at the training scripts and it seems that all other implementations - including momentum sgd - are the same, so I think we should consider the accuracy to be the same.

(3) cudnn exhaustive search: there are 2 reasons not enabling it by default:

(3-1) It seems that NVidia has done a very good job in most recent GPUs, so cudnnGet is usually very close to cudnnFind. My guess is that K80 + new cudnn makes the problem more prominent in this case. For example, I am observing about 47 seconds vs 43 seconds on my Titan X.

(3-2) For many networks in object detection, like @soumith mentioned, doing one cudnnFind and then running that for every combination of size and batch has pretty serious performance regressions. We enabled it in default, but later reverted that decision per @rbgirshick 's experiments and advice. Get gives one a good balance without having to rerun Find every time.

As a result, we actually recommend frameworks to *not* enable it by default especially on new hardware.

(4) A few other things that can help performance: running caffe2/python/memonger.py to reuse memory, and change the network types to dag or async_dag. These are kind of advanced features and we use a mixture of choices for different network types.

(5) In the future, a worthwhile direction is to look at graph rewrite and fusion; for example, NVidia's TensorRT, TensorFlow's XLA, and the TVM projects are good examples along this direction. We are much hopeful it will deliver further performance boosts.

Overall, I think we've learned a lot by looking at your benchmark - especially on how a typical user may use some of the features. We've historically focused on performance and core use cases, and have not invested heavily on tutorials and examples. This is definitely a great data point for us, so thanks again for creating this!
